File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/MicQueueView/PanoMicCollectionCell.m
//
//  PanoMicCollectionCell.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoMicCollectionCell.h"
#import "PanoMicButton.h"
#import <Masonry/Masonry.h>
#import "PanoClientService.h"
#import "UIImage+Extension.h"

@implementation PanoMicCollectionCell
{
    PanoMicButton *micButton;
    
    UILabel *nameLabel;
    
    UIImageView *audioImage;
    
    UIButton *moreButton;
    
    PanoMicInfo *micInfo;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        nameLabel = [UILabel new];
        nameLabel.textColor = [UIColor whiteColor];
        nameLabel.textAlignment = NSTextAlignmentCenter;
        nameLabel.font = [UIFont systemFontOfSize:14];
        nameLabel.text = NSLocalizedString(@"申请上麦", nil);
        [self.contentView addSubview: nameLabel];
        
        micButton = [PanoMicButton buttonWithType:UIButtonTypeCustom];
        [micButton addTarget:self action:@selector(micClickAction) forControlEvents:UIControlEventTouchUpInside];
        [micButton setImage:[UIImage imageNamed:@"room_invite"] forState:UIControlStateNormal];
        [self.contentView  addSubview: micButton];
        
        audioImage = [UIImageView new];
        audioImage.image = [UIImage imageNamed:@"room_audio_mute"];
        audioImage.hidden = true;
        [self.contentView addSubview: audioImage];
        
        moreButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [moreButton addTarget:self action:@selector(moreButtonAction) forControlEvents:UIControlEventTouchUpInside];
        [moreButton setImage:[UIImage imageNamed:@"room_audio_mute"] forState:UIControlStateNormal];
        moreButton.hidden = true;
        [self.contentView  addSubview: moreButton];
        
        [micButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.centerX.mas_equalTo(0);
            make.size.mas_equalTo(CGSizeMake(60, 60));
            make.height.mas_equalTo(micButton.mas_width);
        }];
        
        [audioImage mas_makeConstraints:^(MASConstraintMaker *make) {
            make.size.mas_equalTo(CGSizeMake(20, 20));
            make.top.mas_equalTo(micButton.mas_bottom).offset(-10);
            make.centerX.mas_equalTo(micButton);
        }];
        
        [nameLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.centerX.left.right.mas_equalTo(0);
            make.bottom.mas_equalTo(0).offset(-5);
        }];
        
        [moreButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.top.mas_equalTo(micButton).insets(UIEdgeInsetsMake(20 ,20, 0, 0));
            make.size.mas_equalTo(CGSizeMake(19, 19));
        }];
    }
    return self;
}
    
- (void)micClickAction {
    [_delegate onMicButtonPressed:micInfo];
}
    
- (void)moreButtonAction {
        
}

- (void)updateMicInfo:(PanoMicInfo *)info {
    micInfo = info;
    audioImage.hidden = true;
    moreButton.hidden = true;
    switch (micInfo.status) {
        case PanoMicNone:
            [micButton setImage:[UIImage imageNamed:@"room_invite"] forState:UIControlStateNormal];
            if (PanoClientService.service.userService.isHost) {
                nameLabel.text = NSLocalizedString(@"邀请上麦", nil);
            } else {
                nameLabel.text = NSLocalizedString(@"申请上麦", nil);
            }
            [micButton stopAnimating];
            break;
        case PanoMicConnecting:
        case PanoMicFinished:
        case PanoMicFinishedMuted: {
            if (micInfo.status == PanoMicConnecting) {
                nameLabel.text = NSLocalizedString(@"申请中...", nil);
                [micButton stopAnimating];
            }
            PanoUserInfo *user = info.user;
            if (user)  {
                [micButton setImage:[user avatorImage] forState:UIControlStateNormal];
                [micButton setImage:[user avatorImage] forState:UIControlStateHighlighted];
                if (micInfo.isOnline) {
                    audioImage.hidden = user.audioStatus == PanoUserAudio_Unmute;
                    nameLabel.text = user.userName;
                    if (micInfo.user.isActiving) {
                        [micButton startAnimating];
                    } else {
                        [micButton stopAnimating];
                    }
                }
            }  else {
                [micButton setImage:[UIImage imageNamed:@"room_invite"] forState:UIControlStateNormal];
            }
        }
            break;
    default:
        [micButton setImage:[UIImage imageNamed:@"room_invite"] forState:UIControlStateNormal];
            break;
    }
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/MicQueueView/PanoMicCollectionCell.swift
//
//  PanoMicCollectionCell.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/10/29.
//

import UIKit

protocol PanoMicCollectionCellDelegate : NSObjectProtocol {
    
    func onMicButtonPressed(micInfo: PanoMicInfo);
}


class PanoMicCollectionCell: UICollectionViewCell {
    
    var micButton: PanoMicButton!
    
    var nameLabel: UILabel!
    
    var audioImage: UIImageView!
    
    var moreButton: UIButton!
    
    weak var delegate: PanoMicCollectionCellDelegate?
    
    var micInfo: PanoMicInfo!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        nameLabel = UILabel()
        nameLabel.textColor = UIColor.white
        nameLabel.textAlignment = .center
        nameLabel.font = UIFont.systemFont(ofSize: 14)
        nameLabel.text = NSLocalizedString("申请上麦", comment: "")
        self.contentView.addSubview(nameLabel)
        
        micButton = PanoMicButton(type: .custom)
        micButton.addTarget(self, action: #selector(micClickAction), for: .touchUpInside)
        micButton.setImage(UIImage(named: "room_invite"), for: .normal)
        self.contentView.addSubview(micButton)
        
        audioImage = UIImageView()
        audioImage.image = UIImage(named: "room_audio_mute")
        audioImage.isHidden = true
        self.contentView.addSubview(audioImage)
        
        moreButton = UIButton(type: .custom)
        moreButton.addTarget(self, action: #selector(moreButtonAction), for: .touchUpInside)
        moreButton.setImage(UIImage(named: "room_avator_more"), for: .normal)
        moreButton.isHidden = true
        self.contentView.addSubview(moreButton)
        
        micButton.snp.makeConstraints { (make) in
            make.top.centerX.equalToSuperview()
            make.size.equalTo(CGSize(width: 60, height: 60))
            make.height.equalTo(micButton.snp.width)
        }
        
        audioImage.snp.makeConstraints { (make) in
            make.size.equalTo(CGSize(width: 20, height: 20))
            make.top.equalTo(micButton.snp.bottom).offset(-10)
            make.centerX.equalTo(micButton)
        }
        
        nameLabel.snp.makeConstraints { (make) in
            make.centerX.equalToSuperview()
            make.bottom.equalToSuperview().offset(-5)
            make.left.right.equalToSuperview()
        }
        
        moreButton.snp.makeConstraints { (make) in
            make.left.top.equalTo(micButton).inset(UIEdgeInsets(top: 20 , left: 20, bottom: 0, right: 0))
            make.size.equalTo(CGSize(width: 19, height: 19));
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc func micClickAction() {
        delegate?.onMicButtonPressed(micInfo: micInfo)
    }
    
    @objc func moreButtonAction() {
        
    }
}


extension PanoMicCollectionCell {
    
    func updateMicInfo(info: PanoMicInfo) {
        micInfo = info
//        nameLabel.text = ""
        audioImage.isHidden = true
        moreButton.isHidden = true
        switch micInfo.status {
        case .none?:
            micButton.setImage(UIImage(named: "room_invite"), for: .normal)
            if PanoUserService.isHost() {
                nameLabel.text = NSLocalizedString("邀请上麦", comment: "")
            } else {
                nameLabel.text = NSLocalizedString("申请上麦", comment: "")
            }
            micButton.stopAnimating()
            break
        case .connecting, .finished, .finishedMuted:
            if micInfo.status == .connecting {
                nameLabel.text = NSLocalizedString("申请中...", comment: "")
                micButton.stopAnimating()
            }
            // print("info.user->", info.userId, info.user?.userName)
            if let user = info.user {
                micButton.setImage(user.avatorImage(), for: .normal)
                micButton.setImage(user.avatorImage(), for: .highlighted)
                if micInfo.isOnline() {
                    audioImage.isHidden = user.audioStatus == .unmute
                    nameLabel.text = user.userName
                    if micInfo.user!.isActiving() {
                        micButton.startAnimating()
                    } else {
                        micButton.stopAnimating()
                    }
                }
            } else {
                micButton.setImage(UIImage(named: "room_invite"), for: .normal)
            }
            break
        default:
            micButton.setImage(UIImage(named: "room_invite"), for: .normal)
            break
        }
    }
}
